JENKINS-54662: make UsageStatistics compatible with CasC plugin.

dgarzon opened the pull request:
See [JENKINS-54662](https://issues.jenkins-ci.org/browse/JENKINS-54662).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Entry 1: JENKINS-54662, Make `hudson.model.UsageStatistics` compatible with CasC plugin by introducing `shouldCollect` field.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@daniel-beck 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

Commit:
JENKINS-54662: make UsageStatistics compatible with CasC plugin.

dgarzon commented:
This can be considered a WIP, as it is my first contribution to the project. I would like to get some feedback on the solution I came up with. Thank you!

daniel-beck commented:
Please minimize the whitespace changes. We maintain multiple release lines, and unnecessary changes like this just make that much more cumbersome.

daniel-beck commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -193,18 +202,29 @@ public String getStatData() throws IOException {
+
+    @DataBoundSetter
+    public void setShouldCollectStatistics(boolean shouldCollectStatistics) {
+        this.shouldCollectStatistics = shouldCollectStatistics;
```
This does not write to `Jenkins#setNoUsageStatistics`, so I expect this to not work at all.

Please provide a test that this actually does what you claim.

dgarzon replied:
You are absolutely right! I would have thought the `configure` method gets called, but was wrong. I will move the logic to the `setter` and add tests. Regarding the whitespace changes, it was my IDE doing it on save. I disabled that, and will push without any whitespace changes. Thank you for the review!

dgarzon commented:
@daniel-beck I have added tests to show the functionality now works as expected, and removed all white-space changes from the commit. Feel free to review again. Thank you!

oleg-nenashev reviewed (commented):
The pull request needs to be polished a bit, but I think it can be accepted after the update. Using Descriptor-based DataBoundSetter is obviously a hack to workaround issue with having fields in the `Jenkins` instance (legacy code). My preference in such case would be to go forward with a custom Configurator, but this approach also looks acceptable to me

dgarzon commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -193,13 +205,25 @@ public String getStatData() throws IOException {
+        // for backward compatibility reasons, this configuration is stored in Jenkins
+        Jenkins.get().setNoUsageStatistics(!shouldCollect);
+        this.shouldCollect = shouldCollect;
+        save();
```
@oleg-nenashev I was following the convention set here: https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/REQUIREMENTS.md

dgarzon commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -79,6 +85,12 @@
     private volatile transient long lastAttempt = -1;
 
+    /**
+     * Should we collect statistics given the current configuration?
```
It was meant to be included in the Javadoc. I can change the wording so that it is not a question, but an assertion.

dgarzon commented:
Hi @oleg-nenashev, thank you for the review. I am curious about the implementation you are referring to as "custom configurator". Does that mean removing the field from the `Jenkins` instance, and implementing a new `GlobalConfiguration`? Wouldn't this break backwards compatibility? I would love to work on this, and start contributing to the project. Let me know what you think is the right path forward to get this setting supported by CasC.

oleg-nenashev commented:
@dgarzon This would be the best example: https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/casc/core/AdminWhitelistRuleConfigurator.java . It actually resolves the same problem as you are trying to do here, but on the JCasC side. The advantage of that is that there is no change required on the core's side and hence it can be applied to older core versions.

dgarzon commented:
@oleg-nenashev Nice! I had no idea about this. I also think it is better to have an abstraction for this on the CasC side. Should I go ahead and close this PR, and work on a Configurator on the CasC plugin? I think it is a better solution overall.

oleg-nenashev commented:
@dgarzon my suggestion would be to discuss it with @ewelinawilkosz and @ndeloof in https://gitter.im/jenkinsci/configuration-as-code-plugin

ndeloof reviewed (changes requested):
:-1 on the general approach: don't introduce a new field to replace the one hosted by Jenkins class. Either offer some data migration so this field is actually owned by UsageStatistic, or just introduce the required getter/setters to mimic a legitimate attribute detected by CasC, and delegate to Jenkins
